tensorflow/core/kernels/data/experimental/auto_shard_dataset_op_test.cc

```diff
@@ -0,0 +1,282 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "tensorflow/core/kernels/data/experimental/auto_shard_dataset_op.h"
+
+#include "tensorflow/core/kernels/data/dataset_test_base.h"
+#include "tensorflow/core/kernels/data/shard_dataset_op.h"
+
+namespace tensorflow {
+namespace data {
+namespace {
+
+constexpr char kNodeName[] = "auto_shard_dataset";
+constexpr char kIteratorPrefix[] = "Iterator";
+
+class AutoShardDatasetOpTest : public DatasetOpsTestBase {
+ protected:
+  // Creates a new `AutoShardDataset` op kernel.
+  Status CreateAutoShardDatasetOpKernel(
+      const DataTypeVector& output_types,
+      const std::vector<PartialTensorShape>& output_shapes,
+      std::unique_ptr<OpKernel>* op_kernel) {
+    NodeDef node_def = test::function::NDef(
+        kNodeName, name_utils::OpName(AutoShardDatasetOp::kDatasetType),
+        {AutoShardDatasetOp::kInputDataset, AutoShardDatasetOp::kNumWorkers,
+         AutoShardDatasetOp::kIndex},
+        {{AutoShardDatasetOp::kOutputTypes, output_types},
+         {AutoShardDatasetOp::kOutputShapes, output_shapes}});
+    TF_RETURN_IF_ERROR(CreateOpKernel(node_def, op_kernel));
+    return Status::OK();
+  }
+
+  // Create a new `AutoShardDataset` op kernel context
+  Status CreateAutoShardDatasetContext(
+      OpKernel* const op_kernel,
+      gtl::InlinedVector<TensorValue, 4>* const inputs,
+      std::unique_ptr<OpKernelContext>* context) {
+    TF_RETURN_IF_ERROR(CheckOpKernelInput(*op_kernel, *inputs));
+    TF_RETURN_IF_ERROR(CreateOpKernelContext(op_kernel, inputs, context));
+    return Status::OK();
+  }
+};
+
+struct RangeDatasetParams {
+  int64 start;
+  int64 stop;
+  int64 step;
+};
+
+struct TestCase {
+  RangeDatasetParams range_dataset_param;
+  Tensor num_workers;
+  Tensor index;
+  std::vector<Tensor> expected_outputs;
+  DataTypeVector expected_output_dtypes;
+  std::vector<PartialTensorShape> expected_output_shapes;
+  int64 expected_cardinality;
+  std::vector<int> breakpoints;
+};
+
+// Test Case 1: simple case.
+TestCase TestCase1() {
+  return {/*range_data_param*/ {0, 10, 1},
+          /*num_workers*/
+          DatasetOpsTestBase::CreateTensor<int64>(TensorShape({}), {5}),
```

Comment:
You are writing a lot of repeated code here (e.g. `DatasetOpsTestBase::CreatorTest<int64>...`). Could we add a struct constructor?

```
struct TestCase {
  ...
  Tensor num_workers_,
  ...
  TestCase(int start, int stop, int index, int num_workers, vector<int> expected_outputs, ...) : 
    ..., 
    num_workers_(DatasetOpsTestBase::CreateTensor<int64>(TensorShape({}), {num_workers}), 
    ...
}
```

I think this will make your code simpler to read.

Reply:
Do we need `_` for each member?

Reply:
You are right, we should _not_ have `_` for each member.

Reply:
Thanks for the reference link! 

The struct constructor does make the code clean and simpler to read.